sound/soc/intel/boards/sdw_rt711_rt1308_rt715.c

```diff
@@ -764,61 +768,202 @@ static void set_codec_init_func(const struct snd_soc_acpi_link_adr *link,
 	}
 }
 
-static int create_sdw_codec_dai(struct device *dev,
-				struct snd_soc_dai_link *dai_links, int *id,
-				struct snd_soc_dai_link_component *cpus,
-				const struct snd_soc_acpi_link_adr *link,
-				int *cpus_id)
+/*
+ * check endpoint status in slaves and gather link ID for all slaves in
+ * the same group to generate different CPU DAI. Now only support
+ * one sdw link with all slaves set with only single group id.
+ *
+ * one slave on one sdw link without aggregaged = 0
+ * one sdw link <---> one-cpu DAI <---> one-codec DAI
+ *
+ * two or more slaves on one sdw link without aggregaged = 0:
+ * one sdw link <---> one-cpu DAI <---> multi-codec DAIs
+ *
+ * multiple links with multiple slaves with aggregated = 1:
+ * link N <---> cpu DAIs N <----> codec DAIs I (I >= 1)
+ * link N <---> cpu DAIs M <----> codec DAIs J (J >= 1)
+ * ...
+ */
+static int get_slave_info(const struct snd_soc_acpi_link_adr *adr_link,
+			  int *cpu_dai_id, int *cpu_dai_num,
+			  int *codec_num, int *group_id,
+			  bool *group_generated)
+{
+	const struct snd_soc_acpi_adr_device *adr_d;
+	const struct snd_soc_acpi_link_adr *adr_next;
+	int index = 0;
+
+	*codec_num = adr_link->num_adr;
+	adr_d = adr_link->adr_d;
+
+	/* mask indicates only one sdw link */
+	if (!is_power_of_2(adr_link->mask))
+		return -EINVAL;
+
+	cpu_dai_id[index++] = ffs(adr_link->mask) - 1;
+	if (!adr_d->endpoints->aggregated) {
```

Comment:
`endpoints` is an array, right? It might be more logical to use `adr_d->endpoints[0].aggregated` but just a suggestion, not really important.